src/Arch/X86/X86Rewriter.StringIntrinsics.cs

```diff
@@ -0,0 +1,161 @@
+#region License
+/* 
+ * Copyright (C) 1999-2023 Pavel Tomin.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; see the file COPYING.  If not, write to
+ * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
+ */
+#endregion
+
+using Reko.Core;
+using Reko.Core.Expressions;
+using Reko.Core.Intrinsics;
+using Reko.Core.Types;
+
+namespace Reko.Arch.X86
+{
+    /// <summary>
+    /// Rewrite string intrinsics(strlen, memcpy, etc.).
+    /// </summary>
+    public partial class X86Rewriter
+    {
+        private static readonly IntrinsicProcedure findFirstChange = new IntrinsicBuilder("__find_first_change__", false)
+            .PtrParam(new UnknownType())
+            .PtrParam(new UnknownType())
+            .Returns(new UnknownType());
+
+        private bool RewriteStringIntrinsic()
+        {
+            //$TODO: supported segmented mode.
+            if (arch.ProcessorMode.PointerType == PrimitiveType.SegPtr32)
+                return false;
+            if (!IsAutoIncrementingMode())
+                return false;
+            switch (instrCur.Mnemonic)
+            {
+            case Mnemonic.scasb:
+                // repne scasb
+                if (IsRepNe())
+                    return RewriteScasbIntrinsic();
+                break;
+            case Mnemonic.movs:
+            case Mnemonic.movsb:
+                // rep movs
+                if (IsRep())
+                    return RewriteMovsIntrinsic();
+                break;
+            case Mnemonic.cmps:
+            case Mnemonic.cmpsb:
+                // repe cmps
+                if (IsRepE())
+                    return RewriteCmpsIntrinsic();
```

Comment:
...and `RewriteCmpsToFindFirstDifference`

Reply:
Done in d2c0582